### ISSUE TITLE
Implement or update GitHub workflows to use an up to date workflow file.

### ISSUE THREAD

**Issue by tebb** [localgovdrupal/localgov#176]
Within localgovdrupal there are at least 2 filenames (test.yml, localgov-drupal-ci.yml) and at least 2 versions of test.yml in use in GitHub workflows to conditionally trigger and run tests.

The suggestion is to standardise on one, up to date workflow file that includes recent improvements including use of PHP7.4 and the ability to run tests in forked repos.

@Adnan-cds suggested a "separate issue for test.yml per repo would be better if that's not too much work.".

I have started scripting this.

**Comment by tebb** [comment]
@stephen-cox @andybroomfield @Adnan-cds @finnlewis 

tldr:  Could someone confirm the approach - 1 workfolw yml file runs all tests - and specify the naming of github workflow files project wide please?

Background:

The template for new workflows (needs updating) is in localgovdrupal's .github repo here: https://github.com/localgovdrupal/.github/blob/master/workflow-templates/localgov-drupal-ci.yml

Most repos use test.yml, but there's also localgov-drupal-ci.yml in the template, in localgov_events and in localgov_theme.

localgov_theme's workflow file is different in that it includes a second 'job' using the same triggers for Mocha Javascript tests: https://github.com/localgovdrupal/localgov_theme/blob/master/.github/workflows/localgov-drupal-ci.yml

There may be more non-phpunit testing in future in other repos.

**Comment by finnlewis** [comment]
@paulpopus @Adnan-cds let's discuss this here.

Should we have different configs for testing JS etc?

**Comment by Adnan-cds** [comment]
So to summarize what I said in the standup:
- The Workflow file for most of our repos should use a workflow similar to [our starter template](https://github.com/localgovdrupal/localgov/blob/master/.github/workflows/test.yml).   This helps with future updates.  Repo specific *label changes* are of course perfectly fine.
- localgov_theme is an exception as it has some mocha tests.  These mocha test related workflow config can remain part of the existing workflow file or they can be split into a separate test file.  Both are fine IMHO.
- The test template itself is outdated.  It should be similar to the [localgov repo's Workflow file](https://github.com/localgovdrupal/localgov/blob/master/.github/workflows/test.yml).
- I am happy to stick to the current filename **test.yml** as it clearly expresses the intent.

**Comment by tebb** [comment]
Thanks all for the help.

The pr to update the workflow template as suggested is here: https://github.com/localgovdrupal/.github/pull/7

I believe the updated template, including comments, is repo agnostic, so should support easy, global replacement of existing 'instances' (test.yml) in future, for example when PHP version is increased again.

Mocha tests in the theme are currently a separate 'job' in the single workflow file, which runs in parallel to the PHP unit tests.
  - I haven't spotted any disadvantages to splitting out Mocha tests to a second file
  - Splitting Mocha tests out would support allowing tests.yml to become uniform across repos.

However ...
  - I haven't checked whether multiple workflow files are run in parallel, in a similar way to multiple jobs in the same file.
  - I haven't checked how splitting out would affect the Actions tab.

I will look at those two things unless @paulpopus gets there before me.

If no downside to splitting out, Mocha tests could be in tests-mocha.yml or tests-js.yml or ...

**Comment by paulpopus** [comment]
I've checked on multiple workflow files and by default they do run in parallel and I think that using events you can create dependencies between them.

By splitting the files you get additional entries in the Actions tab but you can filter by workflow specifically. You can see this here https://github.com/localgovdrupal/localgov_theme/actions

I've opted to split the mocha tests from the drupal tests in the theme, I think it makes sense and it should make updating either of them easier in the future. Currently those mocha tests don't rely on Drupal so those tests can be run in parallel.

**Comment by paulpopus** [comment]
Hi @tebb 
I made the assumption that going forward all drupal workflow tests will be named `localgov-drupal-ci.yml` as per the https://github.com/localgovdrupal/.github/tree/master/workflow-templates repo instead of `test.yml`.

Can you confirm which naming convention you prefer?

**Comment by tebb** [comment]
**tldr;**

We currently have 2 and may have more GitHub workflow files.  Personally, I don't mind what we call them, but suggest it makes sense for identical files to have the same name project-wide.  Paul and Adnan have suggested different nameing.  Please advise. 

**Hi @paulpopus**

Yes, I see what you mean - Though the template and 'instance' names don't need to be the same.

Naming convention:  I don't have a strong opinion and as I'm new here, I don't think it's my call :) ... So far I've followed @Adnan-cds suggestion above, to stick with test.yml for existing tests.  If changing, it would be good to decide now.

So we now have identical files with 2 different names: test.yml and localgov-drupal-ci.yml.  It make sense to me, for identical files, even in different repos to have the same name.

I just want agreement, so that I can implement a uniform version of the workflow file for as many repos as possible.  I'm trying to avoid creating noise in the repos later, just to rename a file.

Separate files per 'thing' as you have done in the theme for Mocha tests on the theme make sense to me.

Related:  At the last technical meeting (Were you there at this point?) I understood @stephen-cox to suggest splitting out the PHP coding standards test to get quicker, more easily located, test failures.   I'm not sure if this suggestion is a 3rd file (1:PHP, 2:JS, 3:PHPCS) or a second job in the existing PHP tests file.

The 'localgov-drupal-ci.yml' name possibly came from following the GitHub workflow example.

Consensus would be good ... @finnlewis ... We have ~3 pull requests which include updating (but not renaming) test.yml.  @mattrfletcher and I were holding off with ~3 more awaiting acceptance of the workflow file updates.

**Comment by paulpopus** [comment]
Thanks for getting back, yes I was going to go through all the repos and update the test files to the latest one so that they support but this tiny detail was brought to my attention. The `localgov-drupal-ci.yml` name is more descriptive than `test.yml`

I don't think this is a big deal either way but I'll wait until we have some agreement on things!

**Comment by tebb** [comment]
@paulpopus 

It would probably be easier for you to update numerous repos compared with our fork + pr route - I'm unsure if you commit to master?

It might make sense to be clear what @stephen-cox wants to do re phpcs if that's a straight forward decision, to be able to include that at the same time.

**Comment by stephen-cox** [comment]
At the moment the test script runs 3 separate tests; coding standards and deprecated code checks and the unit tests. I think these should be split into separate tests within the same file. This will make it clearer when one test fails as to what's failed.

I'll create a separate issue for this, but if there's a plan to go through all the repos and update the test file, we might want to wait until this has been done.

I'll take a look this morning.

**Comment by stephen-cox** [comment]
I've created a new issue to discuss splitting tests into separate jobs: https://github.com/localgovdrupal/localgov/issues/178

**Comment by paulpopus** [comment]
Thanks both, I think it makes sense to sort out these tests now, so I'll consolidate the fork changes and the test splits from Stephen's PR and when that's ready and tested on one repo we can do the same for all the rest.

**Comment by stephen-cox** [comment]
Just to update this issue on the progress on splitting tests out into separate jobs, I have a pull request, https://github.com/localgovdrupal/localgov/pull/179, but the PHPUnit tests are failing because it looks like the SQLite database is readonly. I have been trying to find a solution to this, but am not making great progress. Any thoughts or attempts to fix it would be appreciated.

**Comment by stephen-cox** [comment]
I have refactored and renamed the workflow template here: https://github.com/localgovdrupal/.github/pull/9. This has changed the name of the test template file to `test.yml` as this is what's use in most / all of the modules.

**Comment by andybroomfield** [comment]
I think this is done, but re-open if it still needs work.
